log4j-core/src/main/java/org/apache/logging/log4j/core/pattern/ThrowableStackTraceRenderer.java

```diff
@@ -0,0 +1,333 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.logging.log4j.core.pattern;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.jspecify.annotations.NullMarked;
+import org.jspecify.annotations.Nullable;
+
+/**
+ * {@link ThrowableRenderer} implementation for rendering stack traces.
+ *
+ * @param <C> the context type
+ */
+@NullMarked
+class ThrowableStackTraceRenderer<C extends ThrowableStackTraceRenderer.Context> implements ThrowableRenderer {
+
+    private static final RuntimeException MAX_LINE_COUNT_EXCEEDED = new RuntimeException("max-line-count-exceeded");
+
+    private static final String CAUSED_BY_CAPTION = "Caused by: ";
+
+    static final String SUPPRESSED_CAPTION = "Suppressed: ";
+
+    final List<String> ignoredPackageNames;
+
+    final int maxLineCount;
+
+    ThrowableStackTraceRenderer(final List<String> ignoredPackageNames, final int maxLineCount) {
+        this.ignoredPackageNames = ignoredPackageNames;
+        this.maxLineCount = maxLineCount;
+    }
+
+    @Override
+    public final void renderThrowable(
+            final StringBuilder buffer, final Throwable throwable, final String lineSeparator) {
+        if (maxLineCount > 0) {
+            try {
+                C context = createContext(throwable);
+                renderThrowable(buffer, throwable, context, new HashSet<>(), lineSeparator);
+            } catch (final Exception error) {
+                if (error != MAX_LINE_COUNT_EXCEEDED) {
+                    throw error;
+                }
```

Comment:
I am wondering what the performance of throwing `MAX_LINE_COUNT_EXCEEDED` is?
Did you measure the impact of throwing exceptions instead of returning a value?

Reply:
See the added `ThrowVsReturnBenchmark` in beede50bcf252310c4b7f30800ba233a90aa0575:

```
Benchmark   Mode  Cnt     Score    Error  Units
withReturn  avgt   10  1017.747 ± 26.728  ns/op
withThrow   avgt   10  1275.421 ± 24.547  ns/op
```

I think this is acceptable.